filters/playlist.py

```diff
@@ -0,0 +1,176 @@
+from thumbor.loaders import LoaderResult
+
+TOTAL = 4
+CANVAS_WIDTH = 430
+CANVAS_HEIGHT = 249
+CANVAS_BACKGROUND = '#26262d'
+IMAGE_SPEC = [
+    {
+        'x': 5,
+        'y': 5,
+        'w': 165,
+        'h': 239,
+    },
+    {
+        'x': 175,
+        'y': 5,
+        'w': 165,
+        'h': 239,
+    },
+    {
+        'x': 345,
+        'y': 5,
+        'w': 80,
+        'h': 117,
+    },
+    {
+        'x': 345,
+        'y': 127,
+        'w': 80,
+        'h': 117,
+    },
+]
+
+# This code is a variant of filters/distributed_collage.py.
+# usage: http://image_server//unsafe/filters:playlist(img1_url%7Cimg2_url%7Cimg3_url)/img0_url
+
+class Picture:
+    def __init__(self, url, thumbor_filter):
+        self.url = url
+        self.thumbor_filter = thumbor_filter
+        self.extension = splitext(url)[-1].lower()
+        self.engine = None
+        self.fetched = False
+        self.failed = False
+
+    def fill_buffer(self, buffer):
+        if (self.engine is None):
+            self.engine = self.thumbor_filter.create_engine()
+        self.buffer = buffer
+        self.fetched = True
+
+    def request(self):
+        try:
+            self.thumbor_filter.context.modules.loader.load(
+                self.thumbor_filter.context, self.url, self.on_fetch_done)
+        except Exception, err:
+            self.error(err)
+
+    def save_on_disc(self):
+        if self.fetched:
+            try:
+                self.engine.load(self.buffer, self.extension)
+            except Exception, err:
+                self.error(err)
+
+            try:
+                self.thumbor_filter.storage.put(self.url, self.engine.read())
+                self.thumbor_filter.storage.put_crypto(self.url)
+            except Exception, err:
+                self.error(err)
+        else:
+            self.error("Can't save unfetched image")
+
+    def on_fetch_done(self, result):
+        self.fill_buffer(result.buffer if isinstance(result, LoaderResult) else result)
+        self.save_on_disc()
+        self.thumbor_filter.on_image_fetch()
+
+    # Resize image
+    def process(self, width, height):
+        try:
+            self.engine.load(self.buffer, self.extension)
+            self.engine.resize(width, height)
+        except Exception, err:
+            logging.error(err)
+
+    def error(self, msg):
+        logging.error(msg)
+        self.failed = True
+
+# Playlist filter. Stitches source images to generate new image.
+# urls - Ordered list of source image URLs, separated by '|'
+class Filter(BaseFilter):
+    regex = (
+        r'(?:playlist\((?P<urls>[^\)]+)\))'
+    )
+
+    @filter_method(
+        BaseFilter.String,
+        async=True
+    )
+    @tornado.gen.coroutine
+    def playlist(self, callback, urls):
+        logging.debug('playlist filter invoked')
+        self.storage = self.context.modules.storage
+
+        self.callback = callback
+        self.urls = [self.context.request.image_url] + urls.split('|')
```

Comment:
does urls need to be urldecoded?

Reply:
No. In fact, I believe these must remain encoded, based on issues I debugged when developing the 'stitch' filter.